Separate inventory holder info from container & player inventories

dktapps opened the pull request:
<!-- Summarize your PR here. Keep it short and simple. -->
<!-- Explain existing problems or why this pull request is necessary -->
This PR breaks the cyclic dependency between `Inventory` and its holder, which unblocks a lot of new developments.

### Related issues & PRs
- Fixes #5033
- Removes a blocker for #6147 (which in turn means that async tasks will eventually be able to work with tiles)
- Removes a blocker for #2684

## Changes
### API changes
- `Player->getCurrentWindow()` now returns `?InventoryWindow` instead of `?Inventory`
- `Player->setCurrentWindow()` now accepts `?InventoryWindow` instead of `?Inventory`
- `InventoryWindow` introduced, which is created for each player viewing the inventory, provides decorative information like holder info for `InventoryTransactionEvent`, and is destroyed when the window is closed, eliminating cyclic references
- Added:
  - `player\InventoryWindow`
  - `player\PlayerInventoryWindow` - wraps all permanent inventories of Player with type info for transactions
  - `inventory\Hotbar` - replaces all hotbar usages in `PlayerInventory`
  - `Human->getHotbar()`
  - `Human->getMainHandItem()`, `Human->setMainHandItem()`, `Human->getOffHandItem()`, `Human->setOffHandItem()`
  - `block\utils\AnimatedContainerLike` & `block\utils\AnimatedContainerLikeTrait` (for chests, shulkerboxes, etc)
  - `block\utils\Container` & `block\utils\ContainerTrait` for blocks containing items (chests, etc)
  - `block\utils\MenuAccessor` implemented by all blocks that can open inventory menus
  - `block\utils\MenuAccessorTrait` used by blocks with menus but without inventories (anvils, crafting tables etc)
- Removed:
  - `inventory\DelegateInventory` (only used for ender chests)
  - `inventory\PlayerInventory`,
  - `inventory\PlayerOffHandInventory`,
  - `inventory\PlayerCraftingInventory`,
  - `inventory\PlayerCursorInventory` - these have all been internally replaced by `SimpleInventory` & they will appear as `PlayerInventoryWindow` in transactions (check `getType()` against the `PlayerInventoryWindow::TYPE_*` constants to identify them)
  - `block\inventory\AnimatedBlockInventoryTrait`, (blocks now handle this logic directly using `AnimatedContainer` and `AnimatedContainerTrait`)
  - `block\inventory\BlockInventoryTrait`,
  - `block\inventory\BlockInventory`
- Most `BlockInventory` classes have been transitioned to `InventoryWindow` wrappers
- Tiles now all use `SimpleInventory` internally (no cyclic references) except for `Chest` (which uses `CombinedInventory`, without holder info)
- `InventoryOpenEvent` and `InventoryCloseEvent` now provide `InventoryWindow` instead of `Inventory` (to provide type information)
- `InventoryTransaction` and `SlotChangeAction` now provide `InventoryWindow` instead of `Inventory`
- Renamed `TransactionBuilderInventory` to `SlotChangeActionBuilder`
- `TransactionBuilderInventory->getBuilder()` now accepts `InventoryWindow` instead of `Inventory`
- `DoubleChestInventory` superseded by `CombinedInventory` - this new class allows combining any number of inventories behind a single object; mainly used for double chests but plugins could use it to do lots of fun things

### Impacts to plugins
Plugins can now do the following:
```php
$block = $world->getBlockAt($x, $y, $z);
if($block instanceof MenuAccessor){
    $block->openToUnchecked($player);
}
```
As compared to the old way:
```php
$tile = $world->getTileAt($x, $y, $z);
if($tile instanceof Container){
    $player->setCurrentWindow($tile->getInventory());
}
```

#### Advantages
- No tile access needed
- Works for menu blocks without inventories as well as container blocks
- Less code

### Behavioural changes
Inventories no longer keep permanent cyclic references to their holders.

## Backwards compatibility
This makes significant BC breaks. However, all changes are able to be adapted to and the same amount of information is present on all APIs and events.

## Follow-up
- Implement #6147 
- Support inventory inheritance when copying blocks from one position to another

## Tests
Briefly tested in-game.

dktapps committed:
Separate hotbar from player inventory
this allows this functionality to be used with any type of inventory, and also makes it a little nicer to use in many cases.

dktapps committed:
BaseInventory no longer uses viewers to send updates to players
we want viewers to be as close to decorative as possible, so that they provide useful information to plugins, but don't get in the way of other changes.

dktapps committed:
First look: Split up Inventory & InventoryWindow
this unblocks a variety of changes, such as positionless tiles, enhanced APIs on Blocks for inventories, and also eliminates a bunch of cyclic references within the core code.

linked to #5033

dktapps committed:
this bites me every single time

dktapps committed:
Burn more junk

dktapps committed:
Chest block now has responsibility for configuring double chest inventories
it already needs to locate the correct pair anyway to know the left/right for DoubleChestInventoryWindow, so we might as well use this logic for initializing the DoubleChestInventory itself too. The old logic in tile/Chest didn't work correctly.

dktapps committed:
namespace tidy

dktapps committed:
Remove potentially problematic ephemeral window creation
this is *probably* fine, but best avoided.

dktapps committed:
Avoid unnecessary repeated calls

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
CS

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps commented on src/block/inventory/window/EnchantingTableInventoryWindow.php:
```diff
@@ -0,0 +1,110 @@
+		);
+		$this->inventory->getListeners()->add($listener);
+
+		$this->listener = \WeakReference::create($listener);
```
Probably need to clean this up, think I forgot about it

dktapps committed:
Integrate block container animations and SFX into Block classes by way of AnimatedContainer interface
this allows getting rid of several container window classes.

we should probably look into adding more info to the BlockInventoryWindow to make the type easier to identify, though.
now that holder is tracked by an ephemeral window, we can put whatever we like in there.

dktapps committed:
...

dktapps committed:
EnchantingTableInventoryWindow: cleanup weakref mess
this really needs a shorter name

ShockedPlot7560 commented:
I have no real problem with the current implementation. It seems fine to me and is a first step towards a better inventory system.

Another follow-up I would suggest concerns the viewers of the inventory. I can see the point of including viewers in this system but I still have a sticking point because we're including a reference to the player in the inventory interface. Wouldn't it be good after this to think of a system where we have classes that only contain storage logic / listeners and that's it ?   
This may also be a bad idea, potentially making the code into spaghetti for inventories.

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps commented:
> I have no real problem with the current implementation. It seems fine to me and is a first step towards a better inventory system.
> 
> Another follow-up I would suggest concerns the viewers of the inventory. I can see the point of including viewers in this system but I still have a sticking point because we're including a reference to the player in the inventory interface. Wouldn't it be good after this to think of a system where we have classes that only contain storage logic / listeners and that's it ? This may also be a bad idea, potentially making the code into spaghetti for inventories.

Yeah, that's the ideal world. Unfortunately we need the Inventory to still track the viewers so that plugins can see all the viewers of an inventory, and also so that the inventory menus get closed when a block is deleted. I don't see another obvious solution to this problem. We could pull the viewer list out of `Inventory` and have the container itself manage it, but I'm not sure if that would improve anything.

dktapps commented:
Few issues remain with the PR:

- [x] `BlockInventoryWindow` subclasses have minimum 3 constructor parameters - annoying for plugins, lots of room to break things - I'm thinking something like `$chest->openTo(Player $player)` would be a good idea to simplify the usage
- [x] I haven't figured out a solution for stuff like `Furnace` which uses an `InventoryListener` to send itself a scheduled block update on a position. That's a remaining blocker for #6147.
~~Possibly could use `Block` in `BlockInventoryWindow` - this is fine now that the windows aren't referenced by holders anymore, we don't need to worry about cyclic references~~
- [x] No inventory type info in windows for single chest, shulker, barrel & ender chest - this isn't needed for internal use but would probably be desired by plugins; could just let them check the block type but I don't know if this should be bound to blocks

~~Possibly could use `Block` in `BlockInventoryWindow` - this is fine now that the windows aren't referenced by holders anymore, we don't need to worry about cyclic references~~

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
Rename Container(Trait) -> ContainerTile(Trait)
this allows introducing block variations of these without name conflicts

dktapps committed:
Shift inventory management responsibility to World
this removes a bunch of problematic Position usages from Tile, as well as getting rid of a bunch of code duplication.

dktapps committed:
Allow blocks to respond to the contents of their containers being updated
turns out relying on scheduled updates for this was a bad idea, since it causes a lot of unnecessary code to run every tick, as well as being problematic for campfire, which doesn't have any blockstates to compare against.

dktapps commented:
Double chests have also proven to be a pain in the ass and will likely be an obstacle for #6147 as well

dktapps committed:
Remove dodgy code

dktapps committed:
Remove CampfireInventory

dktapps committed:
Campfire block's inventory is now null if it hasn't been set in the world
having this created by the block was unreliable anyway. If items were set into the block's created inventory before setting the block in the world, the campfire contents would get overridden when the block was next run through readStateFromWorld() anyway.

There needs to be a deeper exploration of how to handle blocks with inventories without requiring plugins to interact with tiles. For now, this isn't the worst solution, but it's not the best either.

dktapps committed:
smh

dktapps committed:
Revert "Chest block now has responsibility for configuring double chest inventories"

This reverts commit 1d2b52732e3c475ddc2bab4e45726d22850e3d5c.

I hadn't considered that the likes of plugins and hoppers need to be
able to interact with double chest inventories as well as players.

If we were to move this logic to the Block side, we'd have to expose
APIs on the Chest block to get the correct inventory lazily. I'm not
sure I want to commit to having getInventory() on the block right now,
as we can't guarantee it's available (see problems around Campfire
inventory on the block).

Short term, it'll probably be better to just expose the logic in
block\Chest for deciding which side the inventories should be on.

dktapps committed:
...

dktapps committed:
Replace DoubleChestInventory with a more generic CombinedInventory
this could be used for a bunch of different things aside from double chests
since the DoubleChestInventory no longer references anything specific about chests,
I figured it was time to generalize this.

dktapps committed:
stfu

dktapps committed:
BlockInventoryTrait: include a Block ref instead of Position

dktapps committed:
CombinedInventory now propagates updates if its backing inventories were directly modified
this was always lacking with DoubleChestInventory and is a major factor in it being basically useless for custom use cases.

dktapps committed:
Rename CombinedInventory -> CombinedInventoryProxy

dktapps commented:
Only remaining major issue is around the lengthened usage for opening inventory windows to players.

I didn't quite decide how to deal with it. There are a few things to consider:
- Not all blocks which open windows have inventories of their own (e.g. anvils, ender chests, crafting tables, etc.)
- Blocks which _do_ have inventories of their own may not have those inventories present (e.g. due to missing tiles) and so a mode of allowed failure would need to be introduced (e.g. `Chest->openTo()` would need to be allowed to fail). This sucks, since it will _not_ fail in 99.9% of cases (only those special cases where something hasn't been done correct by a plugin, or old worlds).
- Blocks which have inventories also usually have a feature to test for container lock. The player must use an item with a matching custom name to open such containers. Should `openTo()` check this kind of logic? If not, how do we avoid duplicated logic fetching tile in `onInteract()` vs `openTo()`? (Side note: I suppose plugins already need this boilerplate duplicated logic on stable branches, since we don't offer a simple way to get the inventory anyway.)

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
Update SurvivalBlockBreakHandler.php

dktapps commented:
I think finishing this with nice APIs may require differentiating between "container" and "menu" blocks.
Container blocks have inventories (chest, shulkerbox, hopper, furnace, etc) while menu blocks don't (anvil, crafting table, loom, etc).

Opening a menu block should never fail since these blocks don't have inventories and cannot have locks.
Opening container blocks may fail due to incorrect item key or the tile being missing.

I also think the changes wrt. pushing inventory updates through `World` should probably be reverted also, since these aren't related to the window refactor and I'm also doubtful about the way it's been done. We'll need something like it in the future for tile de-positioning (if that's even actually possible), but for this specific PR it's not required.

dktapps commented:
I'm also wondering whether the player's main inventory window handling should be redone as part of this PR. We'd probably need to introduce a separate "hotbar" PlayerInventoryWindow type to facilitate this.

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -2785,10 +2803,10 @@ public function setCurrentWindow(Inventory $inventory) : bool{
-		$inventoryManager->onCurrentWindowChange($inventory);
-		$inventory->onOpen($this);
-		$this->currentWindow = $inventory;
+		$this->logger->debug("Opening inventory " . get_class($window) . "#" . spl_object_id($window));
```
This can probably be changed in `Opening window` to keep the right vocabulary

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -2797,7 +2815,7 @@ public function removeCurrentWindow() : void{
 		if($this->currentWindow !== null){
 			$currentWindow = $this->currentWindow;
 			$this->logger->debug("Closing inventory " . get_class($this->currentWindow) . "#" . spl_object_id($this->currentWindow));
```
Same here

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -2806,20 +2824,31 @@ public function removeCurrentWindow() : void{
+	protected function addPermanentWindows(array $windows) : void{
+		foreach($windows as $window){
+			$window->onOpen();
+			$this->permanentWindows[spl_object_id($window)] = $window;
```
This can probably now create confusion ? Before, the inventory is normally the same instance while now the window could be different with the same inventory instance.

dktapps replied:
No, these windows are always open, they can't be changed

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
Revert "Remove dodgy code"

This reverts commit 76528b20c11ffcc487ce7be6896fd4e98ccc9e9e.

dktapps committed:
Revert "Allow blocks to respond to the contents of their containers being updated"

This reverts commit 4850bd5538214642b2e96b3f1b4954b1f21c5cdd.

dktapps committed:
Revert "Shift inventory management responsibility to World"

This reverts commit 40574be3336289a813fda403ee858fb372afe067.

dktapps committed:
fix phpstan

dktapps committed:
Blocks are ephemeral objects, InventoryWindow might exceed its lifetime

dktapps committed:
...

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
Merge branch 'major-next' into inventory-rework

dktapps committed:
First step towards standardised opening logic for containers and menu blocks
we'll want to introduce interfaces for these, but getting the code deduplicated is enough to start with.

dktapps committed:
Finally deal with APIs on container and menu blocks
Two main interfaces are introduced:
- Container (chest, furnace, etc)
- MenuAccessor (includes Containers but also anvils, crafting tables etc. which are not containers)

Containers have inventories, locks, and everything from MenuAccessor
MenuAccessors have obstruction checks, and openToUnchecked() for plugins to use

I opted not to include precondition checks in openTo() because it's possible
to replicate those using the provided APIs if desired.

dktapps committed:
Remove dependencies on hotbar and offhand inventory to get equipment
It's always grated at me that the call chain to get the held item is so long

dktapps committed:
Fix build

dktapps committed:
Fix test

dktapps commented:
Alright. I think I'm finally done with this

dktapps committed:
Tidy debug messages

dktapps committed:
tidy

dktapps committed:
oops

dktapps committed:
Abstract getPosition() in traits
we don't technically need to do this, but I suppose it's nicer for people who don't use a static analyser.

dktapps commented:
In the interest of avoiding branch rot I'm going to get this merged, the final result won't look too different from this anyway. Pls let me know if there are issues found after the fact.
